### DISCUSSION TITLE
is there a way for me to get the edge geometry from a Valhalla tile in Python?

### DISCUSSION THREAD

**Discussion by Bondify** [valhalla/valhalla#3796] [Q&A] [answered]
I am using Valhalla's `map-matching` API service to snap GPS points to the street network (which works great by the way!). From the response I get a lot of useful information: the original points matched to the street network, a list of all the `edge_ids` the trip follows, and the complete shape of the trip encoded.

For my specific problem, I would need the actual geometry of each of the edges in the list of all the `edge_ids` I get in the response. I've seen that I could get that geometry using the `locate` API if I try to snap all the points that I would get decoding the shape that I get in the previous point. But I have to clean it afterwards as `locate` responds with some edges that are traversal to the trip.

Is there any way I can just go into the tiles and get the geometry for a given `edge_id` (I get the `tile_id` and `level` using both the Map Matching API and Locate API)

This is my first time asking something here so please let me know if I need to provide more detail.

Thank a lot for your help :)

**Comment by kevinkreiser** [comment, marked as answer]
Yes the python bindings have the full service api and the trace attributes service has begin and end shape indices as options for the output. With these indices you can pull per edge shape from the fully encoded shape after you decode it into an array of lat lons. Look at the attribute filters section of the map matching api docs

**Reply by Bondify** [reply]
I can't believe you replied so quickly @kevinkreiser! this is exactly what I needed, thanks a lot for your help :)

**Reply by nilsnolde** [reply]
Maybe you're already aware of it @Bondify , we also published the Valhalla Python bindings on PyPI earlier this year: https://pypi.org/project/pyvalhalla/

**Reply by Bondify** [reply]
yes, I was aware but haven't tried it yet. Do you think it could be more suitable for my use case as I need to parse the response from json to a pandas DataFrame @nilsnolde ? I haven't dived much (or at all) into the differences between the response I get from calling the Valhalla's APIs with `requests` and using the PyPI binding to be honest.

Thanks again to both for your help

**Reply by nilsnolde** [reply]
Oh it’s giving you the same responses (if the same Valhalla version is used as a base). The advantage is it’s basically using the C++ code directly, not going over HTTP overhead. That saves considerable time, esp if you have shorter traces. Try it, it’s fairly low barrier if you built a Valhalla graph before. If not there’s an easy valhalla docker image you’ll find in our repos as well.

**Reply by Bondify** [reply]
that sounds like a good enough reason for me to try it out, time savings are always welcome and if the response format is the same I wouldn't need to change the wrangling functions. Thanks a lot!
